src/System.Text.JsonLab/System/Text/Json/Utf8JsonReader.cs

```diff
@@ -266,24 +270,61 @@ private void ReadFirstToken(byte first)
             }
             else
             {
-                ConsumeValue(first);
+                _isSingleValue = true;
+                if (ConsumeValue(first))
+                {
+                    if (CurrentIndex >= (uint)_buffer.Length)
+                    {
+                        TokenType = JsonTokenType.Value;
+                        return true;
+                    }
+#if UseInstrumented
```

Comment:
I am not sure I follow. The entire point of the ifdef was to track more info for the actual exception. In most cases, the ifdefs surround codepaths where both readers throw an exception, but one of them is the trigger to re-process the json before throwing. Why would this particular ifdef unique?

Reply:
There are some differences in the readers that are not in error code paths (e.g. incrementing the position, line #). These need to be ifdefed as runtime checks would impact perf. But here, you are already in the error code path. The perf does not matter at that point (throwing exceptions is very slow), and so I don't think we should complicate the code and use ifdefs in error code paths.

Reply:
I addressed this by adding an overload to JsonThrowHelper.ThrowJsonReaderException that accepts Instrumented.Utf8JsonReader, in an attempt to cleanly pass in the custom error message string. Let me know if you have feedback around that approach.

Reply:
I had to revert this change since it caused a significant performance regression (roughly 2x slower across the board). I believe it had to do with passing around a formatted string (which contained property/field values) so the methods were likely not getting optimized as heavily.

Reply:
@AndyAyersMS, do you know what is causing the regression? Is there anything that the JIT can do here or are there tricks/workarounds that you know to reduce the perf hit?

This was the commit that caused a regression: https://github.com/dotnet/corefxlab/pull/2496/commits/c7c45b1893f6be72430c1d065edb4d490380778a

Essentially, moving from calling a basic throw helper: `JsonThrowHelper.ThrowJsonReaderException(ref this);` to something like `JsonThrowHelper.ThrowJsonReaderException(ref this, $"Depth of {Depth} within an object is larger than max depth of {MaxDepth}");`

The assumption was, since the throw helper would only be called in exceptional cases, it should have no negative affect on performance for the common case. That assumption ended up not being true.

Reply:
Have you tried passing a constant string rather than a formatted one?

Reply:
My guess is that the formatting creates enough register pressure that code paths outside this call are impacted. But you'd have to look at profiles or diffs to be sure. If it is really 2x slower it should readily show up in profiling.

If this is the issue, you need to defer formatting. Either pass the format string or pass an enum along with the values needed for message, and do all the formatting in the throw helper(s). This is more or less what the [CoreCLR throw helper](https://github.com/dotnet/coreclr/blob/master/src/System.Private.CoreLib/src/System/ThrowHelper.cs) pattern does.

Reply:
> My guess is that the formatting creates enough register pressure that code paths outside this call are impacted.

That's what it looks like:
**Before vs after comparison:** https://www.diffchecker.com/qo1taY5Q

> ave you tried passing a constant string rather than a formatted one?

That doesn't have any impact on the disassembly (so likely won't cause the regression).

> If this is the issue, you need to defer formatting. Either pass the format string or pass an enum along with the values needed for message, and do all the formatting in the throw helper(s). This is more or less what the CoreCLR throw helper pattern does.

Sounds good, I will give that a shot. Thanks!

When deferring formatting, how do I pass in arbitrary arguments needed for formatting the string properly without a regression? If the throw helper accepts `params` there are extra `call` and `mov` instructions within the caller. Alternatively, I could add several overloads with specific args (results in N move instructions where N is number of args), but only if the jump statement is true.